### ISSUE TITLE
Media queries that contain `calc()` creates erroneous validation error

### ISSUE THREAD

**Issue by pierlon** [ampproject/amphtml#35793]
### Description

Whenever the AMP validator comes across a media query containing an instance `calc()` a validation error is thrown. Here's an example in [AMP Playground](https://playground.amp.dev/#share=PCFET0NUWVBFIGh0bWw+CjxodG1sIOKaoT4KICA8aGVhZD4KICAgIDxtZXRhIGNoYXJzZXQ9InV0Zi04IiAvPgogICAgPHRpdGxlPk15IEFNUCBQYWdlPC90aXRsZT4KICAgIDxsaW5rIHJlbD0iY2Fub25pY2FsIiBocmVmPSJzZWxmLmh0bWwiIC8+CiAgICA8bWV0YSBuYW1lPSJ2aWV3cG9ydCIgY29udGVudD0id2lkdGg9ZGV2aWNlLXdpZHRoIiAvPgogICAgPHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7LW1vei1hbmltYXRpb246LWFtcC1zdGFydCA4cyBzdGVwcygxLGVuZCkgMHMgMSBub3JtYWwgYm90aDstbXMtYW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGg7YW5pbWF0aW9uOi1hbXAtc3RhcnQgOHMgc3RlcHMoMSxlbmQpIDBzIDEgbm9ybWFsIGJvdGh9QC13ZWJraXQta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUAtbW96LWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW1zLWtleWZyYW1lcyAtYW1wLXN0YXJ0e2Zyb217dmlzaWJpbGl0eTpoaWRkZW59dG97dmlzaWJpbGl0eTp2aXNpYmxlfX1ALW8ta2V5ZnJhbWVzIC1hbXAtc3RhcnR7ZnJvbXt2aXNpYmlsaXR5OmhpZGRlbn10b3t2aXNpYmlsaXR5OnZpc2libGV9fUBrZXlmcmFtZXMgLWFtcC1zdGFydHtmcm9te3Zpc2liaWxpdHk6aGlkZGVufXRve3Zpc2liaWxpdHk6dmlzaWJsZX19PC9zdHlsZT48bm9zY3JpcHQ+PHN0eWxlIGFtcC1ib2lsZXJwbGF0ZT5ib2R5ey13ZWJraXQtYW5pbWF0aW9uOm5vbmU7LW1vei1hbmltYXRpb246bm9uZTstbXMtYW5pbWF0aW9uOm5vbmU7YW5pbWF0aW9uOm5vbmV9PC9zdHlsZT48L25vc2NyaXB0PgogICAgPHNjcmlwdCBhc3luYyBzcmM9Imh0dHBzOi8vY2RuLmFtcHByb2plY3Qub3JnL3YwLmpzIj48L3NjcmlwdD4KICAgIDxzdHlsZSBhbXAtY3VzdG9tPgogICAgICBAbWVkaWEgKG1pbi13aWR0aDogY2FsYyg4NDBweCAtIDQ4cHgpKSB7fQogICAgPC9zdHlsZT4KICA8L2hlYWQ+CiAgPGJvZHk+CiAgICA8aDE+SGVsbG8gQU1QSFRNTCBXb3JsZCE8L2gxPgogIDwvYm9keT4KPC9odG1sPgo=) with the media query `@media (min-width: calc(840px - 48px)) {}`:

![image](https://user-images.githubusercontent.com/16200219/130732610-f419ad63-7c5e-4eb6-afa3-e16c2eca698e.png)

That same media query is considered valid by the [W3C CSS validator](http://jigsaw.w3.org/css-validator/#validate_by_input+with_options), however:

<img width="374" alt="image" src="https://user-images.githubusercontent.com/16200219/130732831-95d8299e-3a19-4d43-b711-12739164c650.png">

After doing some debugging I've pinpointed the issue to this block of code where it parses the media feature (in this case `calc(840px - 48px)`):

https://github.com/ampproject/amphtml/blob/7a3474dfea23101d9f039f86a6c5e86be885d66a/validator/js/engine/parse-css.js#L1321-L1338

As you can see it consumes the media feature until it reaches the end of the line or encounters a closing parenthesis, which means it would stop at the last dimension in the media feature (`48px`). Right after [the closing parenthesis is consumed](https://github.com/ampproject/amphtml/blob/7a3474dfea23101d9f039f86a6c5e86be885d66a/validator/js/engine/parse-css.js#L1343) and it does a check to see if it has reached the end of the media query:

https://github.com/ampproject/amphtml/blob/7a3474dfea23101d9f039f86a6c5e86be885d66a/validator/js/engine/parse-css.js#L1229

In this case however the media query is not finished being parsed as the closing parenthesis of the media query has not been consumed, and therefore the validation error is raised:

https://github.com/ampproject/amphtml/blob/7a3474dfea23101d9f039f86a6c5e86be885d66a/validator/js/engine/parse-css.js#L1186-L1189

Now going back to the parsing of the media feature, the [CSS 3 Media Queries spec](https://www.w3.org/TR/css3-mediaqueries/#media1) states:

> Media features only accept single values: one keyword, one number, or a number with a unit identifier. (The only exceptions are the ‘aspect-ratio’ and ‘device-aspect-ratio’ media features.)

That may be misinterpreted as `calc()` not being accepted as a media feature value, but according to the [CSS 3 Values and Units spec](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#calc-notation) it can be used wherever a number or dimension is allowed:

> The `calc()` expression represents the result of the mathematical calculation it contains, using standard operator precedence rules. It can be used wherever [`<length>`](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#length-value), [`<frequency>`](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#frequency-value), [`<angle>`](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#angle-value), [`<time>`](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#time-value), [`<percentage>`](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#percentage-value), [`<number>`](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#number-value), or [`<integer>`](https://www.w3.org/TR/2019/CR-css-values-3-20190606/#integer-value) values are allowed.

---

Not related to the bug, but here are some browser bug issues related to `calc()` and media queries for completeness:

https://bugs.chromium.org/p/chromium/issues/detail?id=421909
https://bugzilla.mozilla.org/show_bug.cgi?id=1256575

### Reproduction Steps

1. Create a file called `test.html` with the following contents:

	  ```html
	  <!doctype html>
	  <html ⚡>
	  <head>
	    <meta charset="utf-8">
	    <link rel="canonical" href="self.html">
	    <meta name="viewport" content="width=device-width,minimum-scale=1">
	    <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
	    <style amp-custom>
	      @media (min-width: calc(840px - 48px)){}
	    </style>
	    <script async src="https://cdn.ampproject.org/v0.js"></script>
	  </head>
	  <body>Hello, AMP world.</body>
	  </html>
	  ```
2. Valiate the HTML file with the AMP JS validator:

```sh
amphtml-validator test.html
```

3. See the following result:

```
test.html: PASS
test.html:9:4 CSS syntax error in tag 'style amp-custom' - malformed media query.
```

### AMP Version Affected

2108132216000

**Comment by Gregable** [comment]
@pierlon First let me start by saying thank you for writing such an incredibly well researched and described issue report.

@honeybadgerdontcare is right, this is only a warning but it's still also a valid bug in the parsing routine. I think it should be very easy to fix though, especially given the analysis you've already done. I'll take it on myself to try to resolve this relatively soon. I believe I was the original author who made the mistake of ignoring calc() in the media expression parsing and taking a shortcut here.

**Comment by Gregable** [comment]
This is now fixed. You may or may not see it in tooling for up to an hour or so while the release goes live.

**Comment by pierlon** [comment]
Thanks @Gregable!

**Comment by westonruter** [comment]
Fixed in #35937.
